File: examples/python/evm/sign_typed_data.py
# Usage: uv run python evm/sign_typed_data.py

import asyncio

from cdp import CdpClient
from cdp.evm_message_types import EIP712Domain


async def main():
    async with CdpClient() as cdp:
        account = await cdp.evm.get_or_create_account(name="MyAccount")
        domain = EIP712Domain(
            name="EIP712Domain",
            chain_id=1,
            verifying_contract="0x0000000000000000000000000000000000000000",
        ).as_dict()
        types = {
            "EIP712Domain": [
                {"name": "name", "type": "string"},
                {"name": "chainId", "type": "uint256"},
                {"name": "verifyingContract", "type": "address"},
            ],
        }
        primary_type = "EIP712Domain"
        message = {
            "name": "EIP712Domain",
            "chainId": 1,
            "verifyingContract": "0x0000000000000000000000000000000000000000",
        }
        signature = await cdp.evm.sign_typed_data(
            address=account.address,
            domain=domain,
            types=types,
            primary_type=primary_type,
            message=message,
        )
        print("Signature: ", signature)


asyncio.run(main())

File: examples/typescript/evm/signTypedData.ts
// Usage: pnpm tsx evm/signTypedData.ts

import { CdpClient } from "@coinbase/cdp-sdk";
import "dotenv/config";

const cdp = new CdpClient();

const account = await cdp.evm.createAccount();

console.log("Created account:", account.address);

const signature = await cdp.evm.signTypedData({
  address: account.address,
  domain: {
    name: "Permit2",
    chainId: 1,
    verifyingContract: "0x000000000022D473030F116dDEE9F6B43aC78BA3",
  },
  types: {
    EIP712Domain: [
      { name: "name", type: "string" },
      { name: "chainId", type: "uint256" },
      { name: "verifyingContract", type: "address" },
    ],
    PermitTransferFrom: [
      { name: "permitted", type: "TokenPermissions" },
      { name: "spender", type: "address" },
      { name: "nonce", type: "uint256" },
      { name: "deadline", type: "uint256" },
    ],
    TokenPermissions: [
      { name: "token", type: "address" },
      { name: "amount", type: "uint256" },
    ],
  },
  primaryType: "PermitTransferFrom",
  message: {
    permitted: {
      token: "0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48",
      amount: "1000000",
    },
    spender: "0xFfFfFfFFfFFfFFfFFfFFFFFffFFFffffFfFFFfFf",
    nonce: "0",
    deadline: "1717123200",
  },
});

console.log("Signature:", signature);
